File: ai-stock-platform/api/services/order_websocket.py
"""
Order WebSocket Service
Created: 2025-05-19 04:49:34
Author: daparthi001
"""
from typing import Dict, Set
from fastapi import WebSocket
import json
import asyncio
from datetime import datetime
from api.models.orders import Order, OrderStatus
from api.utils.websocket_manager import WebSocketManager

class OrderWebSocketService:
    def __init__(self):
        self.active_connections: Dict[str, Set[WebSocket]] = {}
        self.ws_manager = WebSocketManager()

    async def connect(self, websocket: WebSocket, user_id: str):
        """
        Connect a user to the WebSocket
        """
        await self.ws_manager.connect(websocket)
        if user_id not in self.active_connections:
            self.active_connections[user_id] = set()
        self.active_connections[user_id].add(websocket)

    async def disconnect(self, websocket: WebSocket, user_id: str):
        """
        Disconnect a user from the WebSocket
        """
        await self.ws_manager.disconnect(websocket)
        if user_id in self.active_connections:
            self.active_connections[user_id].remove(websocket)
            if not self.active_connections[user_id]:
                del self.active_connections[user_id]

    async def broadcast_order_update(self, user_id: str, order: Order):
        """
        Broadcast order updates to connected clients
        """
        if user_id in self.active_connections:
            message = {
                'type': 'order_update',
                'data': {
                    'order_id': order.id,
                    'symbol': order.symbol,
                    'status': order.status,
                    'executed_quantity': order.executed_quantity,
                    'executed_price': order.executed_price,
                    'timestamp': datetime.utcnow().isoformat()
                }
            }
            
            await self._send_to_user(user_id, message)

    async def broadcast_execution_report(self, user_id: str, execution: Dict):
        """
        Broadcast execution reports to connected clients
        """
        if user_id in self.active_connections:
            message = {
                'type': 'execution_report',
                'data': {
                    'order_id': execution['order_id'],
                    'symbol': execution['symbol'],
                    'executed_quantity': execution['quantity'],
                    'executed_price': execution['price'],
                    'timestamp': execution['timestamp']
                }
            }
            
            await self._send_to_user(user_id, message)

    async def broadcast_order_book_update(self, symbol: str, order_book: Dict):
        """
        Broadcast order book updates to all connected clients
        """
        message = {
            'type': 'order_book_update',
            'symbol': symbol,
            'data': order_book,
            'timestamp': datetime.utcnow().isoformat()
        }
        
        await self.ws_manager.broadcast(message)

    async def _send_to_user(self, user_id: str, message: Dict):
        """
        Send message to all connections of a specific user
        """
        if user_id in self.active_connections:
            for websocket in self.active_connections[user_id]:
                try:
                    await websocket.send_json(message)
                except Exception as e:
                    print(f"Error sending message to websocket: {e}")
                    await self.disconnect(websocket, user_id)

    async def start_heartbeat(self):
        """
        Start heartbeat to keep connections alive
        """
        while True:
            await asyncio.sleep(30)  # Send heartbeat every 30 seconds
            heartbeat = {
                'type': 'heartbeat',
                'timestamp': datetime.utcnow().isoformat()
            }
            await self.ws_manager.broadcast(heartbeat)

File: ai-stock-platform/ui/src/services/OrderWebSocket.ts
/**
 * Order WebSocket Service
 * Created: 2025-05-19 04:54:48
 * Author: daparthi001
 */
import { store } from '../store';
import { updateOrderStatus } from '../store/slices/orderSlice';
import { OrderStatus } from '../types/order';

export class OrderWebSocket {
    private ws: WebSocket | null = null;
    private reconnectAttempts = 0;
    private maxReconnectAttempts = 5;
    private reconnectTimeout = 1000;
    private heartbeatIntervalId: ReturnType<typeof setInterval> | null = null;

    constructor() {
        this.handleMessage = this.handleMessage.bind(this);
        this.handleClose = this.handleClose.bind(this);
        this.handleError = this.handleError.bind(this);
    }

    connect() {
        try {
            const token = localStorage.getItem('token');
            const userId = localStorage.getItem('userId');

            if (!token || !userId) {
                throw new Error('Authentication required');
            }

            const viteEnv: Record<string, string> = (typeof import.meta !== 'undefined' && import.meta.env) || {};
            const wsBase =
                viteEnv.VITE_WS_URL ||
                (typeof process !== 'undefined' ? process.env.REACT_APP_WS_URL : undefined) ||
                'ws://quantumvestai-dev-api:8000/ws';
            this.ws = new WebSocket(`${wsBase}/orders/${userId}?token=${token}`);

            this.ws.onopen = this.handleOpen.bind(this);
            this.ws.onmessage = this.handleMessage;
            this.ws.onclose = this.handleClose;
            this.ws.onerror = this.handleError;

        } catch (error) {
            console.error('WebSocket connection failed:', error);
        }
    }

    disconnect() {
        if (this.ws) {
            this.ws.close();
            this.ws = null;
        }
        if (this.heartbeatIntervalId) {
            clearInterval(this.heartbeatIntervalId);
            this.heartbeatIntervalId = null;
        }
    }

    private handleOpen() {
        console.log('WebSocket connected');
        this.reconnectAttempts = 0;

        // Send heartbeat every 30 seconds
        this.heartbeatIntervalId = setInterval(() => {
            this.sendHeartbeat();
        }, 30000);
    }

    private handleMessage(event: MessageEvent) {
        try {
            const data = JSON.parse(event.data);

            switch (data.type) {
                case 'order_update':
                    this.handleOrderUpdate(data.data);
                    break;
                case 'execution_report':
                    this.handleExecutionReport(data.data);
                    break;
                case 'heartbeat':
                    // Handle heartbeat response
                    break;
                default:
                    console.warn('Unknown message type:', data.type);
            }
        } catch (error) {
            console.error('Error processing WebSocket message:', error);
        }
    }

    private handleClose(event: CloseEvent) {
        console.log('WebSocket closed:', event.code, event.reason);

        if (this.heartbeatIntervalId) {
            clearInterval(this.heartbeatIntervalId);
            this.heartbeatIntervalId = null;
        }

        if (this.reconnectAttempts < this.maxReconnectAttempts) {
            setTimeout(() => {
                this.reconnectAttempts++;
                this.connect();
            }, this.reconnectTimeout * Math.pow(2, this.reconnectAttempts));
        }
    }

    private handleError(error: Event) {
        console.error('WebSocket error:', error);
    }

    private handleOrderUpdate(orderUpdate: any) {
        store.dispatch(updateOrderStatus({
            orderId: orderUpdate.order_id,
            status: orderUpdate.status as OrderStatus,
            executionDetails: {
                executedQuantity: orderUpdate.executed_quantity,
                executedPrice: orderUpdate.executed_price,
                lastUpdateTime: orderUpdate.timestamp
            }
        }));
    }

    private handleExecutionReport(execution: any) {
        store.dispatch(updateOrderStatus({
            orderId: execution.order_id,
            status: 'FILLED' as OrderStatus,
            executionDetails: {
                executedQuantity: execution.quantity,
                executedPrice: execution.price,
                executionTime: execution.timestamp
            }
        }));
    }

    private sendHeartbeat() {
        if (this.ws && this.ws.readyState === WebSocket.OPEN) {
            this.ws.send(JSON.stringify({
                type: 'heartbeat',
                timestamp: new Date().toISOString()
            }));
        }
    }
}